src/lib/api/feed-manip.ts

```diff
@@ -127,168 +167,134 @@ export class FeedViewPostsSlice {
+            // For example, [A ... E -> F], [A ... D -> E], [A ... C -> D], [A -> B -> C]
+            // would get collapsed into [A ... E -> F], with B/C/D considered seen.
+            return false
+          }
+        } else {
+          if (!dryRun) {
+            this.seenUris.add(item.post.uri)
+          }
         }
       }
-    }
-
-    if (!dryRun) {
-      slices = slices.filter(slice => {
-        if (this.seenKeys.has(slice._reactKey)) {
-          return false
-        }
-        for (const item of slice.items) {
-          this.seenUris.add(item.post.uri)
-        }
+      if (!dryRun) {
         this.seenKeys.add(slice._reactKey)
-        return true
-      })
-    }
+      }
+      return true
+    })
 
     return slices
   }
 
   static removeReplies(tuner: FeedTuner, slices: FeedViewPostsSlice[]) {
-    for (let i = slices.length - 1; i >= 0; i--) {
-      if (slices[i].isReply) {
+    for (let i = 0; i < slices.length; i++) {
+      const slice = slices[i]
+      if (
+        slice.isReply &&
+        !slice.isRepost &&
+        // This is not perfect but it's close as we can get to
+        // detecting threads without having to peek ahead.
+        !areSameAuthor(slice.getAllAuthors())
+      ) {
         slices.splice(i, 1)
+        i--
       }
     }
     return slices
   }
 
   static removeReposts(tuner: FeedTuner, slices: FeedViewPostsSlice[]) {
-    for (let i = slices.length - 1; i >= 0; i--) {
+    for (let i = 0; i < slices.length; i++) {
       if (slices[i].isRepost) {
         slices.splice(i, 1)
+        i--
       }
     }
     return slices
   }
 
   static removeQuotePosts(tuner: FeedTuner, slices: FeedViewPostsSlice[]) {
-    for (let i = slices.length - 1; i >= 0; i--) {
+    for (let i = 0; i < slices.length; i++) {
       if (slices[i].isQuotePost) {
         slices.splice(i, 1)
+        i--
       }
     }
     return slices
   }
 
-  static dedupReposts(
+  static removeOrphans(tuner: FeedTuner, slices: FeedViewPostsSlice[]) {
+    for (let i = 0; i < slices.length; i++) {
+      if (slices[i].isOrphan) {
+        slices.splice(i, 1)
+        i--
+      }
+    }
+    return slices
+  }
+
+  static dedupThreads(
     tuner: FeedTuner,
     slices: FeedViewPostsSlice[],
   ): FeedViewPostsSlice[] {
-    // remove duplicates caused by reposts
     for (let i = 0; i < slices.length; i++) {
-      const item1 = slices[i]
-      for (let j = i + 1; j < slices.length; j++) {
-        const item2 = slices[j]
-        if (item2.isThread) {
-          // dont dedup items that are rendering in a thread as this can cause rendering errors
-          continue
-        }
-        if (item1.containsUri(item2.items[0].post.uri)) {
-          slices.splice(j, 1)
-          j--
-        }
+      const rootUri = slices[i].rootUri
+      if (!slices[i].isRepost && tuner.seenRootUris.has(rootUri)) {
+        slices.splice(i, 1)
+        i--
+      } else {
+        tuner.seenRootUris.add(rootUri)
```

Comment:
This is stateful which might be a problem for the "dry run". I'll look if we can remove dry runs as a concept — seems very fragile.